### PR TITLE
Game defined markers, Changes to card properties, big IDE update, and MORE!

### LINKED ISSUE #923: separate markers completely from the set xmls
feel really dumb for thinking of this yesterday, instead of when we did the big revamp of the game definitions.  We kept the actual definitions for the markers attached to the set .xml files, even though they really have no practical attachment to individual sets.  Cuz of this, we still had to stick the marker images in the set GUID folders.  This also added a huge layer of complication when we recently added the image migration to the ImageDatabase folder, as the markers had to stay behind.

So this change would pull markers right out of the set xml file and have them directly controlled by the game def xml.  A <markers> tag would be added, and the markers would be listed almost identically as they were before.  Alternatively (a better option) would be to give the markers its own xml file and set it up like the proxygen def.

Images for the markers would be handled just like other image resources for the game def (such as counters, backgrounds, and default card backs).

`<marker name="Burn Counter" path="/markers/burn.png" id="burn" />`

This would also gut the marker GUIDs entirely, since images would be linked via file paths.

How would python interact?  Instead of requiring the GUID for the marker, we'd just use the new id tag (which can be a word or phrase, or hell keep it as a GUID if you want).

### LINKED ISSUE #1852: Remove "Name" as a pre-set custom card property
Right now a card's name is stored within the custom properties dictionary of its alts.  It really should be its own class property.  Not only would this make the game serializer and IDE simpler, but it would make the code more efficient as currently it has to search the dictionary for the Name custom prop every time it needs to access it.  This is rather silly as the card's Name is a mandatory value, defined in the set XML similarly to its GUID, alt type and size.

There's likely a TON of areas where the card's name is retrieved so this would be a big undertaking.
- [ ] Proxy Gen (template matching and text display)
- [ ] Deck Editor (mainly the card grid)
- [ ] Python API (getting card name)
- [ ] OCTIDE

### PR DESCRIPTION
**Pull request by brine**
This is copy+pasted from my previous pull request, so the newer stuff is probably missing.  I kept it in mostly so it hooks up all the tagged issues again.

> Game-defined markers. Closes #923. Markers are now defined in the game XML, instead of in a set XML. Some modifications were made to the networking code to allow markers to be indexed by a string ID instead of GUIDs. There is full backwards compatibility, as GameSerializer will import any markers found in the sets into the game data. There are no functional changes to the python or play table implementation -- GUIDs can continue to be used as identifiers as strings if preferred. o8build will block compiling of games that have set-defined markers, but the XML schema and gameserializer will allow them for compatibility.
> 
> Cards have a dedicated Name property in its Alternate type, instead of the Name being a custom property added alongside all the other game-defined custom properties.
> 
> The logic behind these changes is to make the card name a dedicated property in the card PropertySet. Previously card name was stored as a custom property alongside all the other game-defined custom properties. Which is a bit of an oddity considering the card Name is a mandatory attribute in the set XMLs (in the same category as Size, Type (for alts), and Id) (closes #1852)
> 
> This required a pretty substantial overhaul in all the various places in the code that use the card's name. While I was doing this, I decided to also clean up how OCTGN deserializes custom properties, especially with inheritance of the base property for alternates. This new version no longer uses the isUndefined property in the PropertyDef objects. Instead, when a card is missing one or more custom properties, the gameserializer simply skips over that property. Some extension methods were added to deal with inheritance and searching for missing properties.
> 
> HUGE OCTIDE upgrades. With this PR, the IDE is nearly functional. We are still missing a python script editor, asset manager and to hook up o8build for package compiling.
> 
> Backend efficiency updates to gameserializer and OCTGN data entities.
> 
> Moved booster pack generator code into PackExtensionMethods
> 
> maybe more??
> 
> This may require extensive testing and proof-checking, to make sure I didn't miss any crucial areas of the code.
> 
>

### CODE REVIEW

File: octgnFX/Octgn/Play/Marker.cs [hunk 1]
```diff
@@ -18,61 +18,61 @@ public class Marker : INotifyPropertyChanged
                                                                                                       new Guid(0, 0, 0,
                                                                                                                0, 0, 0,
                                                                                                                0, 0, 0,
-                                                                                                               0, 1)),
+                                                                                                               0, 1).ToString()),
```
Comment (BenMatteson) [inline octgnFX/Octgn/Play/Marker.cs:21]: I guess this is less likely to conflict than a real name, except the part where that's a terrible guid :P

File: octgnFX/Octgn.Core/DataExtensionMethods/GameExtensionMethods.cs [hunk 2]
```diff
@@ -193,26 +200,19 @@ public static DataTable ToDataTable(this IEnumerable<Card> cards, Game game)
...
-                            values[0] = prop.Value;
-                            continue;
-                        }
-                        var ix = indexes.Where(x => x.Value == prop.Key.Name).Select(x => new { Key = x.Key, Value = x.Value }).FirstOrDefault();
+                        var ix = indexes.Where(x => x.Value == prop.Key.Name).Select(x => new { x.Key, x.Value }).FirstOrDefault();
                         if (ix == null)
                             throw new UserMessageException(L.D.Exception__CanNotCreateDeckMissingCardProperty);
                         if (prop.Key.Type == PropertyType.Integer)
                         {
                             int garbo;
```
[the remark this answers is not in the document]
Comment (BenMatteson) [reply octgnFX/Octgn.Core/DataExtensionMethods/GameExtensionMethods.cs:214]: how did this end up below the other? I guess it goes in order...
so with _above_

File: OCTGN.sln.licenseheader [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//  * License, v. 2.0. If a copy of the MPL was not distributed with this
+//  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extensions: .aspx .ascx
+<%-- 
+Copyright (c) 2011 rubicon IT GmbH
+--%>
+extensions: .vb
+'Sample license text.
+extensions:  .xml .config .xsd .xaml
```
[the remark this answers is not in the document]
Comment (brine) [reply OCTGN.sln.licenseheader:13]: I think this cuts off the xml / xaml extension in line 13?

[the remark this answers is not in the document]
Comment (BenMatteson) [reply OCTGN.sln.licenseheader:13]: shit, I missed, took me like 10 tries to get multi-line suggestions to work in the first place, guess I got the wrong lines when it finally worked

File: OCTGN.sln.licenseheader [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+// /* This Source Code Form is subject to the terms of the Mozilla Public
+//  * License, v. 2.0. If a copy of the MPL was not distributed with this
+//  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extensions: .aspx .ascx
+<%-- 
+Copyright (c) 2011 rubicon IT GmbH
+--%>
+extensions: .vb
+'Sample license text.
```
Comment (BenMatteson) [inline OCTGN.sln.licenseheader:12]:
```suggestion
```
Take two

File: octgnFX/Octgn/DeckBuilder/SearchControl.xaml.cs [hunk 5]
```diff
@@ -57,18 +58,18 @@ public SearchControl(DataNew.Entities.Game game, DeckBuilderWindow deckWindow)
...
+
+            filtersList.ItemsSource = source;
             GenerateColumns(game);
             //resultsGrid.ItemsSource = game.SelectCards(null).DefaultView;
             UpdateDataGrid(game.AllCards(true).ToDataTable(Game).DefaultView);
             FileName = "";
             UpdateCount();
-        }//Why are we populating the list on load? I'd rather wait until the search is run with no parameters (V)_V
-        // Why are we poluting the code with snide comments instead of fixing the problem or making a generic TODO (V)_V
-        // Actually, the more that I think about it, the more I think that the first comment is actually a bad idea. (V)_V
```
[the remark this answers is not in the document]
Comment (brine) [reply octgnFX/Octgn/DeckBuilder/SearchControl.xaml.cs:71]: I know its annoying but is this still a valid issue with the search control?

[the remark this answers is not in the document]
Comment (brine) [reply octgnFX/Octgn/DeckBuilder/SearchControl.xaml.cs:71]: oh wait I removed it already... gosh this review system is confusing.  Do we want to keep this?

File: octgnFX/Octgn/Play/Marker.cs (hunk 1 above)
Comment (brine) [reply octgnFX/Octgn/Play/Marker.cs:21]: yeah I'm open to changing this when we merge in the validation code into gameserializer, that way we can actually reserve names for these default markers.

File: octgnFX/Octgn/DeckBuilder/SearchControl.xaml.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenMatteson) [reply octgnFX/Octgn/DeckBuilder/SearchControl.xaml.cs:71]: nah, I was just letting them be obnoxious one last time

File: octgnFX/Octgn.DataNew/Entities/OptionsList.cs [hunk 6]
```diff
@@ -1,62 +1,14 @@
 ﻿namespace Octgn.DataNew.Entities
 {
-    using System;
-    using System.Security.Cryptography;
     using System.Collections.Generic;
-    using System.Globalization;
-    using System.Xml;
 
     public class OptionsList : IPackItem
```
[the remark this answers is not in the document]
Comment (brine) [reply octgnFX/Octgn.DataNew/Entities/OptionsList.cs]: I can go either way with this one.  Technically the <options> xml element can hold attributes, which would need to be set as properties here.  If we clear it out, we'd need to go back and undo the changes if we decided in the future to make use of this element

**Comment by brine** [comment]
OK this should address some newly discovered concerns with the branch

**Pull request merged**